Setup.makeEnvironment throws an exception if setup is incomplete

This can be reproduced by creating an empty setup directory:

``` shell
$ rm -rf .local/share/libisabelle/setups/Isabelle2015/
$ mkdir -p .local/share/libisabelle/setups/Isabelle2015/
```

Now, running this program

``` scala
val version = Version("2015")
val setup = Setup.defaultSetup(version)
val env = setup.flatMap(_.makeEnvironment)
Await.result(env, Duration.Inf)
```

yields the following exception

```
java.lang.reflect.InvocationTargetException
        at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
        at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
        at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
        at java.lang.reflect.Constructor.newInstance(Constructor.java:422)
        at edu.tum.cs.isabelle.Implementations$.makeEnvironment(Implementations.scala:40)
        at edu.tum.cs.isabelle.Implementations$$anonfun$makeEnvironment$2.apply(Implementations.scala:110)
        at edu.tum.cs.isabelle.Implementations$$anonfun$makeEnvironment$2.apply(Implementations.scala:110)
        at scala.Option.map(Option.scala:146)
        at edu.tum.cs.isabelle.Implementations.makeEnvironment(Implementations.scala:110)
        at edu.tum.cs.isabelle.setup.Setup$$anonfun$makeEnvironment$1.apply(Setup.scala:171)
        at edu.tum.cs.isabelle.setup.Setup$$anonfun$makeEnvironment$1.apply(Setup.scala:169)
        at scala.util.Success$$anonfun$map$1.apply(Try.scala:237)
        at scala.util.Try$.apply(Try.scala:192)
        at scala.util.Success.map(Try.scala:237)
        at scala.concurrent.Future$$anonfun$map$1.apply(Future.scala:235)
        at scala.concurrent.Future$$anonfun$map$1.apply(Future.scala:235)
        at scala.concurrent.impl.CallbackRunnable.run(Promise.scala:32)
        at scala.concurrent.impl.ExecutionContextImpl$AdaptedForkJoinTask.exec(ExecutionContextImpl.scala:121)
        at scala.concurrent.forkjoin.ForkJoinTask.doExec(ForkJoinTask.java:260)
        at scala.concurrent.forkjoin.ForkJoinPool$WorkQueue.pollAndExecAll(ForkJoinPool.java:1253)
        at scala.concurrent.forkjoin.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1346)
        at scala.concurrent.forkjoin.ForkJoinPool.runWorker(ForkJoinPool.java:1979)
        at scala.concurrent.forkjoin.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:107)
Caused by: java.io.IOException: Cannot run program "/home/frank/.local/share/libisabelle/setups/Isabelle2015/bin/isabelle": error=2, No such file or directory
        at java.lang.ProcessBuilder.start(ProcessBuilder.java:1048)
        at isabelle.Isabelle_System$.raw_execute(isabelle_system.scala:284)
        at isabelle.Isabelle_System$.init(isabelle_system.scala:122)
        at edu.tum.cs.isabelle.impl.Environment.<init>(Environment.scala:13)
        at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
        at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
        at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
        at java.lang.reflect.Constructor.newInstance(Constructor.java:422)
        at edu.tum.cs.isabelle.Implementations$.makeEnvironment(Implementations.scala:40)
        at edu.tum.cs.isabelle.Implementations$$anonfun$makeEnvironment$2.apply(Implementations.scala:110)
        at edu.tum.cs.isabelle.Implementations$$anonfun$makeEnvironment$2.apply(Implementations.scala:110)
        at scala.Option.map(Option.scala:146)
        at edu.tum.cs.isabelle.Implementations.makeEnvironment(Implementations.scala:110)
        at edu.tum.cs.isabelle.setup.Setup$$anonfun$makeEnvironment$1.apply(Setup.scala:171)
        at edu.tum.cs.isabelle.setup.Setup$$anonfun$makeEnvironment$1.apply(Setup.scala:169)
        at scala.util.Success$$anonfun$map$1.apply(Try.scala:237)
        at scala.util.Try$.apply(Try.scala:192)
        at scala.util.Success.map(Try.scala:237)
        at scala.concurrent.Future$$anonfun$map$1.apply(Future.scala:235)
        at scala.concurrent.Future$$anonfun$map$1.apply(Future.scala:235)
        at scala.concurrent.impl.CallbackRunnable.run(Promise.scala:32)
        at scala.concurrent.impl.ExecutionContextImpl$AdaptedForkJoinTask.exec(ExecutionContextImpl.scala:121)
        at scala.concurrent.forkjoin.ForkJoinTask.doExec(ForkJoinTask.java:260)
        at scala.concurrent.forkjoin.ForkJoinPool$WorkQueue.pollAndExecAll(ForkJoinPool.java:1253)
        at scala.concurrent.forkjoin.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1346)
        at scala.concurrent.forkjoin.ForkJoinPool.runWorker(ForkJoinPool.java:1979)
        at scala.concurrent.forkjoin.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:107)
Caused by: java.io.IOException: error=2, No such file or directory
        at java.lang.UNIXProcess.forkAndExec(Native Method)
        at java.lang.UNIXProcess.<init>(UNIXProcess.java:248)
        at java.lang.ProcessImpl.start(ProcessImpl.java:134)
        at java.lang.ProcessBuilder.start(ProcessBuilder.java:1029)
        at isabelle.Isabelle_System$.raw_execute(isabelle_system.scala:284)
        at isabelle.Isabelle_System$.init(isabelle_system.scala:122)
        at edu.tum.cs.isabelle.impl.Environment.<init>(Environment.scala:13)
        at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
        at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
        at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
        at java.lang.reflect.Constructor.newInstance(Constructor.java:422)
        at edu.tum.cs.isabelle.Implementations$.makeEnvironment(Implementations.scala:40)
        at edu.tum.cs.isabelle.Implementations$$anonfun$makeEnvironment$2.apply(Implementations.scala:110)
        at edu.tum.cs.isabelle.Implementations$$anonfun$makeEnvironment$2.apply(Implementations.scala:110)
        at scala.Option.map(Option.scala:146)
        at edu.tum.cs.isabelle.Implementations.makeEnvironment(Implementations.scala:110)
        at edu.tum.cs.isabelle.setup.Setup$$anonfun$makeEnvironment$1.apply(Setup.scala:171)
        at edu.tum.cs.isabelle.setup.Setup$$anonfun$makeEnvironment$1.apply(Setup.scala:169)
        at scala.util.Success$$anonfun$map$1.apply(Try.scala:237)
        at scala.util.Try$.apply(Try.scala:192)
        at scala.util.Success.map(Try.scala:237)
        at scala.concurrent.Future$$anonfun$map$1.apply(Future.scala:235)
        at scala.concurrent.Future$$anonfun$map$1.apply(Future.scala:235)
        at scala.concurrent.impl.CallbackRunnable.run(Promise.scala:32)
        at scala.concurrent.impl.ExecutionContextImpl$AdaptedForkJoinTask.exec(ExecutionContextImpl.scala:121)
        at scala.concurrent.forkjoin.ForkJoinTask.doExec(ForkJoinTask.java:260)
        at scala.concurrent.forkjoin.ForkJoinPool$WorkQueue.pollAndExecAll(ForkJoinPool.java:1253)
        at scala.concurrent.forkjoin.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1346)
        at scala.concurrent.forkjoin.ForkJoinPool.runWorker(ForkJoinPool.java:1979)
        at scala.concurrent.forkjoin.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:107)
```

Removing the Isabelle2015 directory and running the above program again fixes the problem.

It seems that this happened to me because the first download of Isabelle2015 was incomplete (there was only an empty `ANNOUNCE` file). It would be nice if libisabelle would check if a setup is complete before running `bin/isabelle`.

--- a/cli/src/main/scala/Main.scala
+++ b/cli/src/main/scala/Main.scala
@@ -8,6 +8,8 @@ import scala.concurrent.ExecutionContext.Implicits.global
 
 import org.log4s._
 
+import cats.data.Xor
+
 import edu.tum.cs.isabelle.api._
 import edu.tum.cs.isabelle.setup._
 
@@ -101,7 +103,11 @@ object Main {
       logger.info(s"Using $configuration")
 
       def mkSetup = args.home match {
-        case None => Setup.defaultSetup(version)
+        case None =>
+          Setup.defaultSetup(version) match {
+            case Xor.Right(setup) => setup
+            case Xor.Left(reason) => sys.error(reason.explain)
+          }
         case Some(home) => Future.successful(Setup(home, guessPlatform, version, Setup.defaultPackageName))
       }
 

--- a/libisabelle/src/main/scala/System.scala
+++ b/libisabelle/src/main/scala/System.scala
@@ -11,6 +11,8 @@ import acyclic.file
 /** Functions to build and create [[System systems]]. */
 object System {
 
+  class StartupException() extends RuntimeException("System startup failed")
+
   /**
    * Synchronously build a
    * [[edu.tum.cs.isabelle.api.Configuration configuration]].
@@ -96,9 +98,8 @@ object System {
    * Build products will be read from `~/.isabelle` on the file system.
    */
   def create(env: Environment, config: Configuration): Future[System] = {
-    val started = Promise[Unit]
-    val system = new System(env, config, { () => started.success(()); () })
-    started.future.map(_ => system)(env.executionContext)
+    val system = new System(env, config)
+    system.initPromise.future.map(_ => system)(env.executionContext)
   }
 
 }
@@ -111,7 +112,7 @@ object System {
  *
  * @see [[edu.tum.cs.isabelle.setup.Setup]]
  */
-final class System private(val env: Environment, config: Configuration, callback: () => Unit) {
+final class System private(val env: Environment, config: Configuration) {
 
   /**
    * The [[scala.concurrent.ExecutionContext execution context]] used
@@ -140,8 +141,14 @@ final class System private(val env: Environment, config: Configuration, callback
   @volatile private var pending = Map.empty[Long, System.OperationState[_]]
 
   private def consumer(markup: Markup, body: XML.Body): Unit = (markup, body) match {
-    case ((env.initTag, _), _) => initPromise.success(()); ()
-    case ((env.exitTag, _), _) => exitPromise.success(()); ()
+    case ((env.initTag, _), _) =>
+      env.sendOptions(session)
+      initPromise.success(())
+      ()
+    case ((env.exitTag, _), _) =>
+      initPromise.tryFailure(new System.StartupException())
+      exitPromise.success(())
+      ()
     case _ =>
       synchronized {
         pending =
@@ -152,11 +159,6 @@ final class System private(val env: Environment, config: Configuration, callback
   }
 
   private val session = env.create(config, consumer)
-
-  initPromise.future foreach { _ =>
-    env.sendOptions(session)
-    callback()
-  }
 
   /**
    * Instruct the prover to shutdown.

--- a/setup/src/main/scala/Platform.scala
+++ b/setup/src/main/scala/Platform.scala
@@ -1,10 +1,15 @@
 package edu.tum.cs.isabelle.setup
 
 import java.net.URL
-import java.nio.channels.FileChannel
+import java.nio.channels.{FileChannel, FileLock}
 import java.nio.file._
 
+import scala.concurrent.{ExecutionContext, Future}
+import scala.util.control.NonFatal
+
 import org.apache.commons.lang3.SystemUtils
+
+import org.log4s._
 
 import edu.tum.cs.isabelle.api.{BuildInfo, Version}
 
@@ -62,6 +67,8 @@ object Platform {
  */
 sealed abstract class Platform {
 
+  private val logger = getLogger
+
   def localStorage: Path
 
   final def setupStorage: Path =
@@ -76,20 +83,35 @@ sealed abstract class Platform {
   final def lockFile: Path =
     localStorage.resolve(".lock")
 
-  def withLock[A](f: () => A): A = {
+  private def acquireLock(): Option[FileLock] = {
     Files.createDirectories(localStorage)
-    Option(FileChannel.open(lockFile, StandardOpenOption.CREATE, StandardOpenOption.WRITE).tryLock()) match {
-      case None =>
-        sys.error("lock could not be acquired")
-      case Some(lock) =>
-        try {
-          f()
-        }
-        finally {
-          lock.close()
-        }
+    FileChannel.open(lockFile, StandardOpenOption.CREATE, StandardOpenOption.WRITE).tryLock() match {
+      case null =>
+        logger.warn("lock could not be acquired")
+        None
+      case lock =>
+        Some(lock)
     }
   }
+
+  def withLock[A](f: () => A): Option[A] = acquireLock().map { lock =>
+    try {
+      f()
+    }
+    finally {
+      lock.close()
+    }
+  }
+
+  def withAsyncLock[A](f: () => Future[A])(implicit ec: ExecutionContext): Future[A] = acquireLock() match {
+    case None =>
+      Future.failed(new RuntimeException("lock could not be acquired"))
+    case Some(lock) =>
+      f().map { a =>
+        try { lock.close() } catch { case NonFatal(_) => }
+        a
+      }
+    }
 
 }
 
@@ -106,11 +128,10 @@ sealed abstract class OfficialPlatform private[isabelle](val name: String) exten
     s"https://isabelle.in.tum.de/website-Isabelle${version.identifier}/dist/Isabelle${version.identifier}"
 
   /**
-   * Internet location containing an archive of the requested
-   * [[edu.tum.cs.isabelle.api.Version version]] for this platform, if
-   * available.
+   * HTTP location containing an archive of the requested
+   * [[edu.tum.cs.isabelle.api.Version version]] for this platform.
    */
-  def url(version: Version): Option[URL] =
-    Some(new URL(s"${baseURL(version)}_$name.tar.gz"))
+  def url(version: Version): URL =
+    new URL(s"${baseURL(version)}_$name.tar.gz")
 
 }

--- a/setup/src/main/scala/Setup.scala
+++ b/setup/src/main/scala/Setup.scala
@@ -5,6 +5,9 @@ import java.net.{URL, URLClassLoader}
 import java.nio.file.{Files, Path, Paths}
 
 import scala.concurrent.{Future, ExecutionContext}
+import scala.util._
+
+import cats.data.Xor
 
 import coursier._
 
@@ -24,62 +27,84 @@ import acyclic.file
  */
 object Setup {
 
+  sealed trait NoSetupReason
+  sealed trait SetupImpossibleReason { def explain: String }
+  case object Absent extends NoSetupReason
+  case class Corrupted(path: Path) extends NoSetupReason with SetupImpossibleReason {
+    def explain = s"Possibly corrupted setup detected at $path; try deleting that folder and running again"
+  }
+  case class Busy(path: Path) extends NoSetupReason with SetupImpossibleReason {
+    def explain = s"File lock $path could not be acquired (busy)"
+  }
+  case object UnknownPlatform extends SetupImpossibleReason {
+    def explain = "Impossible to download setup on unknown platform"
+  }
+
   private val logger = getLogger
 
   /** Default platform: [[Platform.guess guessing]]. */
-  def defaultPlatform: Option[Platform] = {
-    val guess = Platform.guess
-    logger.debug(
-      guess match {
-        case Some(p) => s"Using default platform; detected $p"
-        case None    => "Using default platform; platform could not be detected"
-      }
-    )
-    guess
-  }
+  def defaultPlatform: Option[OfficialPlatform] = Platform.guess
 
+  /**
+   * Default package name of PIDE jars.
+   *
+   * @see [[edu.tum.cs.isabelle.api.Environment]]
+   */
   val defaultPackageName: String = "edu.tum.cs.isabelle.impl"
 
-  // FIXME this whole thing needs proper error handling
+  /**
+   * Location of the success marker file.
+   *
+   * Detection of setups works by looking for the success marker file in the
+   * path of the setup. If the root path of the setup is present, but not the
+   * file, the setup is considered corrupted, for example because of a partial
+   * download.
+   */
+  def successMarker(path: Path): Path =
+    path.resolve(".success")
 
-  def install(platform: OfficialPlatform, version: Version)(implicit ec: ExecutionContext): Future[Setup] =
-    platform.url(version) match {
-      case None =>
-        sys.error("couldn't determine URL")
-      case Some(url) =>
-        logger.debug(s"Downloading setup $version to ${platform.setupStorage}")
-        val stream = Tar.download(url)
-        Files.createDirectories(platform.setupStorage)
-        platform.withLock { () =>
-          Tar.extractTo(platform.setupStorage, stream).map(Setup(_, platform, version, defaultPackageName))
+  def install(platform: OfficialPlatform, version: Version)(implicit ec: ExecutionContext): Future[Setup] = {
+    Files.createDirectories(platform.setupStorage)
+    val url = platform.url(version)
+    logger.debug(s"Downloading setup $version from $url to ${platform.setupStorage}")
+    Tar.download(url) match {
+      case Success(stream) =>
+        val future = platform.withAsyncLock { () =>
+          Tar.extractTo(platform.setupStorage, stream).map { path =>
+            Files.createFile(successMarker(path))
+            Setup(path, platform, version, defaultPackageName)
+          }
         }
-    }
-
-  def detectSetup(platform: Platform, version: Version): Option[Setup] = platform.withLock { () =>
-    val path = platform.setupStorage(version)
-    if (Files.isDirectory(path)) {
-      logger.debug(s"Using default setup; detected $version at $path")
-      Some(Setup(path, platform, version, defaultPackageName))
-    }
-    else {
-      logger.debug(s"Using default setup; no setup found in ${platform.setupStorage}")
-      None
+        future.onComplete { _ => stream.close() }
+        future
+      case Failure(ex) =>
+        logger.error(ex)(s"Failed to download $url")
+        Future.failed(ex)
     }
   }
 
-  def defaultSetup(version: Version)(implicit ec: ExecutionContext): Future[Setup] =
+  def detectSetup(platform: Platform, version: Version): Xor[NoSetupReason, Setup] = platform.withLock { () =>
+    val path = platform.setupStorage(version)
+    if (Files.isDirectory(path)) {
+      if (Files.isRegularFile(successMarker(path)))
+        Xor.right(Setup(path, platform, version, defaultPackageName))
+      else
+        Xor.left(Corrupted(path))
+    }
+    else
+      Xor.left(Absent)
+  }.getOrElse(Xor.left(Busy(platform.lockFile)))
+
+  def defaultSetup(version: Version)(implicit ec: ExecutionContext): Xor[SetupImpossibleReason, Future[Setup]] =
     defaultPlatform match {
       case None =>
-        sys.error("couldn't determine platform")
+        Xor.left(UnknownPlatform)
       case Some(platform) =>
         detectSetup(platform, version) match {
-          case Some(install) =>
-            Future.successful(install)
-          case None =>
-            platform match {
-              case o: OfficialPlatform => install(o, version)
-              case _ => sys.error("unofficial platform can't be automatically installed")
-            }
+          case Xor.Right(install) =>     Xor.right(Future.successful(install))
+          case Xor.Left(Absent) =>       Xor.right(install(platform, version))
+          case Xor.Left(Busy(p)) =>      Xor.left(Busy(p))
+          case Xor.Left(Corrupted(p)) => Xor.left(Corrupted(p))
         }
     }
 
@@ -131,7 +156,7 @@ object Setup {
     def fetchArtifact(artifact: Artifact, cachePolicy: CachePolicy) =
       Cache.file(artifact, logger = Some(downloadLogger), cachePolicy = cachePolicy)
 
-    platform.withLock { () =>
+    platform.withAsyncLock { () =>
       for {
         i <- resolve("interface")
         v <- resolve(version.identifier)

--- a/setup/src/main/scala/Tar.scala
+++ b/setup/src/main/scala/Tar.scala
@@ -5,6 +5,7 @@ import java.net.URL
 import java.nio.file._
 
 import scala.concurrent._
+import scala.util.Try
 
 import org.apache.commons.compress.archivers.tar.{TarArchiveEntry, TarArchiveInputStream}
 import org.apache.commons.compress.compressors.gzip.GzipCompressorInputStream
@@ -17,13 +18,11 @@ import acyclic.file
 /** Downloading and unpacking `tar` archives. */
 object Tar {
 
-  def download(url: URL): TarArchiveInputStream =
-    new TarArchiveInputStream(new GzipCompressorInputStream(url.openStream()))
+  def download(url: URL): Try[TarArchiveInputStream] =
+    Try(new TarArchiveInputStream(new GzipCompressorInputStream(url.openStream())))
 
   def extractTo(path: Path, tar: TarArchiveInputStream)(implicit ec: ExecutionContext): Future[Path] = Future {
-    def next() = blocking {
-      Option(tar.getNextTarEntry())
-    }
+    def next() = Option(tar.getNextTarEntry())
 
     @annotation.tailrec
     def go(entry: Option[TarArchiveEntry], paths: List[Path]): List[Path] = entry match {
@@ -41,12 +40,11 @@ object Tar {
             Files.createSymbolicLink(subpath, Paths.get(entry.getLinkName))
           else if (entry.isLink)
             Files.createLink(subpath, path.resolve(Paths.get(entry.getLinkName)))
-          else if (entry.isFile)
-            blocking {
-              Files.copy(tar, subpath)
-              if (!SystemUtils.IS_OS_WINDOWS)
-                MoreFiles.setMode(subpath, entry.getMode)
-            }
+          else if (entry.isFile) {
+            Files.copy(tar, subpath)
+            if (!SystemUtils.IS_OS_WINDOWS)
+              MoreFiles.setMode(subpath, entry.getMode)
+          }
           else
             sys.error("unknown tar file entry")
         }
@@ -58,7 +56,7 @@ object Tar {
         go(next(), p ::: paths)
     }
 
-    go(next(), Nil).foldLeft(List.empty[Path]) { (roots, path) =>
+    blocking { go(next(), Nil) }.foldLeft(List.empty[Path]) { (roots, path) =>
       if (roots.exists(path.startsWith))
         roots
       else

--- a/setup/src/main/scala/japi/JSetup.scala
+++ b/setup/src/main/scala/japi/JSetup.scala
@@ -6,13 +6,20 @@ import scala.concurrent.{Await, Future}
 import scala.concurrent.ExecutionContext.Implicits.global
 import scala.concurrent.duration.Duration
 
+import cats.data.Xor
+
 import edu.tum.cs.isabelle.api._
 import edu.tum.cs.isabelle.setup._
 
 object JSetup {
 
   def defaultSetup(version: Version, timeout: Duration): Setup =
-    Await.result(Setup.defaultSetup(version), timeout)
+    Setup.defaultSetup(version) match {
+      case Xor.Right(future) =>
+        Await.result(future, timeout)
+      case Xor.Left(reason) =>
+        sys.error(reason.explain)
+    }
 
   def defaultSetup(version: Version): Setup =
     defaultSetup(version, Duration.Inf)

--- a/tests/src/test/isabelle/ROOT
+++ b/tests/src/test/isabelle/ROOT
@@ -1,0 +1,3 @@
+session Unbuilt_Session = Protocol +
+  theories
+    Sleepy

--- a/tests/src/test/scala/DefaultSetup.scala
+++ b/tests/src/test/scala/DefaultSetup.scala
@@ -20,7 +20,7 @@ trait DefaultSetup { self: Specification =>
     ).map(Version.apply).get
 
   lazy val platform: Platform = Setup.defaultPlatform.get
-  lazy val setup: Setup = Setup.detectSetup(platform, version).get
+  lazy val setup: Setup = Setup.detectSetup(platform, version).getOrElse(sys.error("no setup"))
   lazy val env: Future[Environment] = setup.makeEnvironment
   lazy val resources: Resources = Resources.dumpIsabelleResources()
   lazy val config: Configuration = resources.makeConfiguration(Nil, "Protocol")

--- a/tests/src/test/scala/SetupSpec.scala
+++ b/tests/src/test/scala/SetupSpec.scala
@@ -1,0 +1,36 @@
+package edu.tum.cs.isabelle.tests
+
+import java.nio.file.Files
+
+import org.specs2.Specification
+import org.specs2.specification.core.Env
+
+import cats.data.Xor
+
+import edu.tum.cs.isabelle.setup._
+
+class SetupSpec(val specs2Env: Env) extends Specification with DefaultSetup { def is = s2"""
+
+  Isabelle setup
+
+  Isabelle setup detection
+    should handle absent setups      $absent
+    should handle corrupted setups   $corrupted"""
+
+  def absent = {
+    val dir = Files.createTempDirectory("libisabelle_test")
+    val platform = Platform.genericPlatform(dir)
+    Files.createDirectories(platform.setupStorage)
+    Setup.detectSetup(platform, version).toEither must beLeft(Setup.Absent)
+  }
+
+  def corrupted = {
+    val dir = Files.createTempDirectory("libisabelle_test")
+    val platform = Platform.genericPlatform(dir)
+    Files.createDirectories(platform.setupStorage(version))
+    Setup.detectSetup(platform, version).toEither must beLeft.like {
+      case Setup.Corrupted(_) => true
+    }
+  }
+
+}

--- a/tests/src/test/scala/SystemSpec.scala
+++ b/tests/src/test/scala/SystemSpec.scala
@@ -1,0 +1,29 @@
+package edu.tum.cs.isabelle.tests
+
+import java.nio.file.Paths
+
+import scala.concurrent._
+import scala.concurrent.duration._
+import scala.math.BigInt
+
+import org.specs2.Specification
+import org.specs2.specification.core.Env
+
+import edu.tum.cs.isabelle._
+import edu.tum.cs.isabelle.api._
+import edu.tum.cs.isabelle.pure._
+import edu.tum.cs.isabelle.hol._
+
+class SystemSpec(val specs2Env: Env) extends Specification with DefaultSetup with IsabelleMatchers { def is = s2"""
+
+  Isabelle system
+
+  An Isabelle system
+    detects startup failures   ${failingSystem.failed must beAnInstanceOf[System.StartupException].awaitFor(timeout)}"""
+
+
+  def timeout = 10.seconds
+
+  val failingSystem = env.flatMap(System.create(_, resources.makeConfiguration(Nil, "Unbuilt_Session")))
+
+}

